hudi-client/src/main/java/org/apache/hudi/metadata/HoodieMetadataImpl.java

```diff
@@ -0,0 +1,1003 @@
+    // the metadata table was created. We should not sync any instants before that creation time.
+    Option<HoodieInstant> oldestMetaInstant = Option.empty();
+    if (!compactionTimestamp.isPresent()) {
+      oldestMetaInstant = Option.fromJavaOptional(metaTimeline.getInstants()
+          .filter(i -> i.isCompleted() && i.getAction().equals(HoodieTimeline.DELTA_COMMIT_ACTION))
+          .findFirst());
+      if (oldestMetaInstant.isPresent()) {
+        // TODO: Ensure this is the instant at which we created the metadata table
+      }
+    }
+
+    String metaSyncTimestamp = compactionTimestamp.isPresent() ? compactionTimestamp.get()
+        : oldestMetaInstant.isPresent() ? oldestMetaInstant.get().getTimestamp() : "";
+
+    // Metadata table is updated when an instant is completed except for the following:
+    //  CLEAN: metadata table is updated during inflight. So for CLEAN we accept inflight actions.
+    List<HoodieInstant> datasetInstants = datasetMetaClient.getActiveTimeline().getInstants()
+        .filter(i -> i.isCompleted() || (i.getAction().equals(HoodieTimeline.CLEAN_ACTION) && i.isInflight()))
+        .filter(i -> metaSyncTimestamp.isEmpty()
+            || HoodieTimeline.compareTimestamps(i.getTimestamp(), HoodieTimeline.GREATER_THAN_OR_EQUALS,
+                metaSyncTimestamp))
+        .collect(Collectors.toList());
+
+    // Each operation on dataset leads to a delta-commit on the metadata MOR table. So find only delta-commit
+    // instants in metadata table which are after the last compaction.
+    Map<String, HoodieInstant> metadataInstantMap = metaTimeline.getInstants()
+        .filter(i -> i.isCompleted() && i.getAction().equals(HoodieTimeline.DELTA_COMMIT_ACTION)
+            && (metaSyncTimestamp.isEmpty() || HoodieTimeline.compareTimestamps(i.getTimestamp(),
+                HoodieTimeline.GREATER_THAN_OR_EQUALS,
+                metaSyncTimestamp)))
+        .collect(Collectors.toMap(HoodieInstant::getTimestamp, Function.identity()));
+
+    List<HoodieInstant> instantsToSync = new LinkedList<>();
+    datasetInstants.forEach(instant -> {
+      if (metadataInstantMap.containsKey(instant.getTimestamp())) {
+        // instant already synced to metadata table
+        if (!instantsToSync.isEmpty()) {
+          LOG.warn("Found out-of-order already synced instant " + instant + ". Instants to sync=" + instantsToSync);
+        }
+      } else {
+        instantsToSync.add(instant);
+      }
+    });
+
+    // TODO: async clean and async compaction are not handled here. They have a timestamp which is in the past
+    // (when the operation was scheduled) and even on completion they retain their old timestamp.
+    return instantsToSync;
+  }
+
+  /**
+   * Update from {@code HoodieCommitMetadata}.
+   *
+   * @param commitMetadata {@code HoodieCommitMetadata}
+   * @param instantTime Timestamp at which the commit was performed
+   */
+  void update(HoodieCommitMetadata commitMetadata, String instantTime) {
```

Comment:
I feel these update methods can be moved to a new class, which encapsulates all code for reading/writing to file system metadata.